go/internal/cbor/deterministic.go

```diff
@@ -0,0 +1,121 @@
+// byte sequences and checks that they follow the deterministic principles described here:
+// https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c.
+func Deterministic(input []byte) bool {
+	index := 0
+
+	for index < len(input) {
+		// TODO(sonkkeli): Remove debug print.
+		fmt.Printf("index=%v\n", index)
+
+		length, err := deterministicRec(input[index:])
+		if err != nil {
+			fmt.Println(err)
+			return false
+		}
+		index += length + 1
+	}
+	return true
+}
+
+// deterministicRec is a recursively called helper function to check the deterministicy of a bytes array.
+// TODO(sonkkeli): Recursively called parts will be the maps and arrays, which contain other types, meaning
+// unsigned integers, byte strings, text strings.
+func deterministicRec(input []byte) (int, error) {
+	cborItemLengthInBytes := 0
+
+	b := input[0]
+
+	switch getMajorType(b) {
+	case TypePosInt:
+		length, err := unsignedIntegerDeterministic(input)
+		if err != nil {
+			return 0, err
+		}
+		cborItemLengthInBytes = length
+
+		break
+
+	case TypeBytes, TypeText:
+		// TODO(sonkkeli):
+		// length, err := textOrBytesDeterministic(input[index:])
+		break
+
+	case TypeArray:
+		// TODO(sonkkeli):
+		// length, err := arrayDeterministic(input[index:])
+		break
+
+	case TypeMap:
+		// TODO(sonkkeli):
+		// length, err := mapDeterministic(input[index:])
+		break
+
+	default:
+		return 0, errors.New("Missing implementation for major type.")
+	}
+
+	// TODO(sonkkeli): Remove debug print.
+	fmt.Printf("cborItemLengthInBytes=%v\n", cborItemLengthInBytes)
+
+	return cborItemLengthInBytes, nil
+
+}
+
+// unsignedIntegerDeterministic calculates the value of the unsigned integer to ensure that
+// the right amount of bytes is used in the CBOR encoding and returns the length of the unsigned integer.
+func unsignedIntegerDeterministic(input []byte) (int, error) {
+	ainfo := convertToAdditionalInfo(input[0])
+
+	if ainfo == AdditionalInfoIndefinite || ainfo == AdditionalInfoReserved {
+		return -1, fmt.Errorf("%v should not be in use in deterministic CBOR.", ainfo)
+	}
+
+	lengthInBytes, err := ainfo.getAdditionalInfoLength()
+	if err != nil {
+		return -1, err
+	}
+
+	_, err = unsignedIntegerCompareValueToLimit(input[0 : lengthInBytes+1])
+	if err != nil {
+		return -1, err
+	}
+
+	return lengthInBytes, nil
+}
+
+// unsignedIntegerCompareValueToLimit compares the value with the amount of bytes used to ensure that
+// the right amount of bytes is used for the CBOR encoded unsigned integer.
+func unsignedIntegerCompareValueToLimit(input []byte) (uint64, error) {
+	ainfo := convertToAdditionalInfo(input[0])
+
+	limit, err := ainfo.getAdditionalInfoValueLowerLimit()
+	if err != nil {
+		return 0, err
+	}
+
+	value := getUnsignedIntegerValue(input[0:])
+
+	if value < limit {
+		return 0, fmt.Errorf("When following CBOR deterministic principles, %v should not be represented with %v bytes.", value, len(input)-1)
+	}
+	return value, nil
```

Comment:
These functions and their splitting may seem a bit odd because there are some functionality still coming up. I could remove this since it's not needed right now, but it will be needed when the map and array come to the picture. (Because we want to check that the number representing the amount of items on the array/map is also deterministic)

Reply:
Acknowledged.